Fix caps, keyboard group logic

tvquizphd opened the pull request:
**Update**: For the latest status, see [my latest comment](https://github.com/wooorm/nspell/pull/39#issuecomment-760574350).

This closes issue #37, closes issue #41, and ignores issue #46.

This addresses [a `retext-spell` issue](https://github.com/retextjs/retext-spell/issues/20) I've discovered that is fundamentally [an `nspell` issue](https://github.com/wooorm/nspell/issues/37).

<!---
I'm including a concrete example of the problem for reference.

Differences from `main` are **bolded**:

#### One Example I've tested

- assume the input word is "Twinz"
- assume character `a`is "t" at index 0

1. "T" is uppercase in "Twinz".
2. The first group is "qwertzuop".
    1. **We ignore the position of "t" in the group.**
    2. We split the word "Twinz" at **index 0 into `before`="" and `after` = "winz"**
    3. We iterate thrice to group character `b`="e" 
        - We select an uppercase "E" for **index 0** because "T" was uppercase at index 0

This is how we arrive at the reasonable suggestion of **"Ewinz"**.

**Here is how the same logic leads to the correct solution:**

- assume the input word is "Twinz"
- assume character `a`is **"z" at index 4**

1. **"z" is lowercase** in "Twinz".
2. The first group is "qwertzuop".
    1. **We ignore the position of "z" in the group.**
    2. We split the word "Twinz" at index 4 into `before`="Twin" and `after` = ""
    3. We iterate thrice to group character `b`="e" 
        - We select a **lowercase** "e" for index 4 **because "z" was lowercase at index 4.**

This is how we arrive at the correct suggestion of **"Twine"**.
-->

## Addressing a bigger problem

In `main`, the notion of "Keyboard groups" doesn't really make sense as it replaces at a position (of a character in group) that is unrelated to the original position (of a character in word). **This solution attempts to address that "root problem."**

~~For another solution that ignores issue #41 but solves issue #37, see [my alternative PR](https://github.com/wooorm/nspell/pull/38).~~

This solution **splits the word at the original index** of the character in the input word instead of the unrelated position of the character in the keyboard group. See my review comments.

## It's ready to merge!

tvquizphd committed:
Split at character index rather than group position when inserting keyboard edits

tvquizphd commented on lib/suggest.js:
```diff
@@ -77,6 +77,8 @@ function suggest(value) {
   while (++index < length) {
     character = value.charAt(index)
+    before = value.slice(0, index)
+    after = value.slice(index + 1)
```
Here we `slice` the word at the index of the `character` in the input word. From my perspective, this more accurately simulates keyboard errors than the current approach.

tvquizphd commented on lib/suggest.js:
```diff
@@ -90,8 +92,6 @@ function suggest(value) {
       }
 
-      before = value.slice(0, position)
-      after = value.slice(position + 1)
```
Why does the current approach `slice` the word at the position of the `character` in an unrelated `group` string such as the keyboard group "qwertzuop"?

tvquizphd reviewed (commented):
This change has the effect that only upper-case replacements are suggested to replace upper-case characters, and only lower-case replacements are suggested to replace lower-case characters.

tvquizphd commented:
The only test that fails in a way that does not simply involve a reordering of expected values is [test #46](https://travis-ci.org/github/wooorm/nspell/jobs/752620410#L341). The other PR actually [only fails #46](https://travis-ci.org/github/wooorm/nspell/jobs/752620374#L321), but it more closely matches the expected value.

tvquizphd committed:
Disallow duplicate replacements for each index

tvquizphd committed:
replicate original keyboard group limitations

tvquizphd commented:
I've modified this PR significantly in commit [76c6be3](https://github.com/wooorm/nspell/pull/39/commits/76c6be3ba47eeae1830e96c145df2bbf830f19df)  to reproduce the buggy behavior of `nspell/main` in order the match all test cases except test #46. Before we accept this PR, I'd either like to revert commit 76c6be3 entirely (and break the tests) or to rework the while loops to reduce the complexity of "keyboard" edit creation from `O(value.length * groups.length)` to `O(min(value.length, group.length) * groups.length)`.

wooorm commented:
Nice! I think the new suggestions (`[ 'Brandi', 'Shandy', 'handy', 'Ghana', 'Grand', 'Grandee', 'Grands', 'Shanxi', 'hand', 'hands' ]` for `Ghandi`) might be better than what was suggested before btw, I’m fine with changing that one

tvquizphd committed:
ignore original keyboard group limitations

tvquizphd commented:
I've reworked this PR to be more efficient, look more like `main`, and keep the exact behavior of the tests @wooorm looked at yesterday. Those changes have been committed to this PR, but there are a couple changes I'm hesitant to commit.

**To frame the decision that I remains**: I have opened [a PR against this PR on my own fork](https://github.com/tvquizphd/nspell/pull/1). It should be very quick to look at the **4 lines** that PR changes in order to get a sense of why I believe that PR makes more logical sense, but will cause more breaking changes for anyone who relies on the exact ordering of the results. 

I plan to take a look at the two options using a suite of test cases I had originally developed for the capitalization issue (issue #37) in order to get a general sense of how the two options differ.

In practice, it may not matter much whether the keyboard groups behave "as expected" or "as they have been".

wooorm commented:
Nice, thanks for all the research!

Your extra changes on the PR to the PR also look good. I do wonder how much they’d change the tests: from your explanation in the comment above I assume they’d make more sense, in which case, I’m :+1: on it

tvquizphd committed:
update existing tests to reflect new behavior

tvquizphd committed:
Make capitalization more consistent and add new capitalization tests

tvquizphd commented on test/index.js:
```diff
@@ -257,13 +257,24 @@ test('NSpell()', function (t) {
     st.deepEqual(
       us.suggest('Ghandi'),
-      ['shandy', 'Brandi', 'Ghana', 'Grand', 'Grandee', 'Grands', 'handy'],
+      [
+        'Brandi',
+        'Ghana',
+        'Grandee',
+        'Hand',
+        'Hands',
+        'Handy',
+        'Hanoi',
+        'Hindi',
+        'Randi',
+        'Shanxi'
+      ],
       'should suggest alternatives'
```
The old results favored "shandy" because of two edits of "shandi" and "Ghandy", which we lose because there's no group with "s" and "g" or with "i" and "y". The old results contained "Grand" and "Grands" because of an edit of "Grandi", which we lose because there's no group with "r" and "h".

The suggestions of 'Hand', 'Hands', 'Handy', 'Hanoi', 'Hindi', 'Randi' all come from the truncation of "Ghandi" to "Handi" at index 63 of `firstValues`. It is the only non-edit for consideration in the second call to `generate`. Since the old version coincidentally had a net of 1 more "edit" than the updated version, "Handi" was pushed to index 64 of `firstValues` and thus missed the `size` cutoff.

tvquizphd commented on test/index.js:
```diff
@@ -301,7 +312,7 @@ test('NSpell()', function (t) {
     st.deepEqual(
       us.suggest('collor'),
-      ['color', 'collar', 'colloq'],
+      ['color', 'colloq', 'collar'],
       'should suggest removals'
```
Here we favor "colloq" because "r" and "q" share a keyboard group while "a" and "o" do not.

tvquizphd commented on test/index.js:
```diff
@@ -478,27 +489,39 @@ test('NSpell()', function (t) {
+    )
+
+    st.deepEqual(
+      us.suggest('Acnada'),
+      ['Canada'],
+      'should suggest sentence-case with swapped first character'
```
This test actually fails on main with "CANADA" because all-caps is considered a valid spelling while "cAnada" is not considered valid.

tvquizphd commented on test/index.js:
```diff
@@ -478,27 +489,39 @@ test('NSpell()', function (t) {
-        'dent',
         'cont',
-        'font',
-        'wont',
+        'dent',
         'dint',
+        'done',
+        'dong',
+        'font',
```
For "dont" we now actually have 5 equally-weighted keyboard edits instead of the previous sole edit of "dent".

tvquizphd commented on test/index.js:
```diff
@@ -478,27 +489,39 @@ test('NSpell()', function (t) {
     )
 
+    st.deepEqual(
+      us.suggest('Iffect'),
+      ['Affect', 'Effect', 'Infect'],
+      'should suggest sentence-case with replaced first character'
```
This test actually fails on main with "affect", "effect", and "Infect", as the edits were not capitalized correctly.

tvquizphd committed:
update contraction test to reflect new behavior

tvquizphd commented on test/index.js:
```diff
@@ -478,27 +489,45 @@ test('NSpell()', function (t) {
+
+    st.deepEqual(
+      us.suggest('COLORFU'),
+      ['COLORFUL'],
+      'should suggest alternatives for upper-case with added letter'
```
This test fails on main with "colorful" because `upper` was measured to be false even for the end of an all-caps string.

tvquizphd commented on lib/suggest.js:
```diff
@@ -318,14 +336,19 @@ function generate(context, memory, words, edits) {
-
         // Try upper-case if the original character was upper-cased.
         if (upper && inject !== inject.toUpperCase()) {
+          if (currentCase !== 's') {
+            check(before + inject + after)
+            check(before + inject + nextAfter)
+          }
```
This avoids double weighting of lowercase words when both lowercase and sentence case are tested for the same word.

Without this line, the "dont" test would artificially favor results such as "font" over results such as "dent" because "font" would be counted twice.

tvquizphd commented on lib/suggest.js:
```diff
@@ -302,7 +304,23 @@ function generate(context, memory, words, edits) {
+      }
+
+      if (nextAfter && upper !== nextUpper) {
+        // Remove.
+        check(before + switchCase(nextAfter))
+
+        // Switch.
+        check(
+          before +
+            switchCase(nextCharacter) +
+            switchCase(character) +
+            nextNextAfter
+        )
```
These remove and switch in case-specific ways when the case of the current character differs from the case of the next character.

Without this extra logic, the "Ghandi" test would fail by missing "Hanoi", "Hindi" and "Randi" and by returning "hand", "hands", and "handy" in lowercase.

tvquizphd commented:
So I found a few miscellaneous capitalization issues with main that I thought I'd address. I'm operating under a few assumptions I've made by observing the general behavior of `nspell/main`:

- UPPERCASE INPUT should always produce UPPERCASE OUTPUT
- Sentence Case Input should produce Sentence Case Output
   - unless the output is in the dictionary as UPPERCASE
- lowercase input should output lowercase
   - unless the output is in the dictionary as Sentence Case or UPPERCASE 
- FuNkY caSE should output FuNkY caSE
   - unless the output is in the dictionary as Sentence Case or UPPERCASE

tvquizphd committed:
adjust behavior for funky case when sentence-case in dictionary

tvquizphd commented on test/index.js:
```diff
@@ -522,6 +522,12 @@ test('NSpell()', function (t) {
     )
 
+    st.deepEqual(
+      us.suggest('abDUL'),
+      ['Abdul'],
+      'should suggest sentence-case for funky case when sentence-case in dictionary'
```
On main, this test fails with "ABDUL" instead of "Abdul".

tvquizphd commented:
I just realized that I have unnecessarily conflated things by adding the changes to capitalization as well as the original fix to the keyboard edits. I can split these out into separate PRs to make things clearer...

tvquizphd commented:
So, to close issue #37 and issue #41, I've set up 3 options for @wooorm to decide:
- Merge this PR or PR #45. They'd have the same result, but #45 is less messy.
- Merge PR #42. This **only changes the keyboard edits**.
  - I could open a new issue for the other capitalization changes, or we could just ignore them.

tvquizphd commented:
> I do wonder how much they’d change the tests: from your explanation in the comment above I assume they’d make more sense, in which case, I’m 👍 on it

There are essentially two options for how the "Ghandi" case should be handled:

- Look at PR #42 results: ['Brandi', 'Ghana', 'Grandee', 'Shanxi', 'hand', 'hands', 'handy'](https://github.com/wooorm/nspell/pull/42/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR266).
    - Three results are based on a **lowercase truncation** of "Ghandi" to "handi".
    - This is without the changes to capitalization I've snuck into this PR
- Look at this PR's [results](https://github.com/wooorm/nspell/pull/39/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR266) or identically PR #45 [results](https://github.com/wooorm/nspell/pull/45/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR266):
    -   'Brandi', 'Ghana', 'Grandee', 'Hand', 'Hands', 'Handy', 'Hanoi', 'Hindi', 'Randi', 'Shanxi'
    - Four results depend on the **uppercase truncation** of "Ghandi" to "Handi".

tvquizphd commented:
I'd just like to clarify that 

PR #42 and PR #45 are both **ready to merge**.

tvquizphd committed:
Simplify this PR by reverting all additional capitalization fixes

tvquizphd commented:
After a bit of thought, I think the new issue #46 is off-topic for this PR, so I've reverted changes that address that issue.

**To make this PR easier to follow**, I've collapsed any off-topic/outdated comments I've made.

This PR now only addresses issues #37 and #41, as originally intended.

tvquizphd committed:
add more sensible test to ensure 100% coverage

tvquizphd commented:
I realize the new issue #46 is off-topic for this PR, so I've reverted changes that address that issue. To make this PR easier to follow, I've collapsed any off-topic/outdated comments I've made. This PR now only addresses issues #37 and #41, as originally intended.

Now it's easiest to understand this PR by looking at the files changed. Despite all the commits I've added to this PR, it only alters lib/suggest by +11/-2 lines and test/index by +13/-7 lines.

wooorm commented:
Awesome, thanks!

wooorm commented:
released in [2.1.5](https://github.com/wooorm/nspell/releases/tag/2.1.5)!
